sale_warehouse_calendar/tests/test_sale_order_dates.py

```diff
@@ -0,0 +1,124 @@
+    post_install = True
+
+    @classmethod
+    def _define_calendar(cls, name, attendances):
+        return cls.env["resource.calendar"].create(
+            {
+                "name": name,
+                "attendance_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "name": "%s_%d" % (name, index),
+                            "hour_from": att[0],
+                            "hour_to": att[1],
+                            "dayofweek": str(att[2]),
+                        },
+                    )
+                    for index, att in enumerate(attendances)
+                ],
+            }
+        )
+
+    @classmethod
+    def setupClassCompany(cls):
+        cls.company = cls.env.user.company_id
+        cls.company.security_lead = 1
+
+    @classmethod
+    def setUpClassCalendar(cls):
+        cls.calendar = cls._define_calendar("40 Hours", [(8, 16, i) for i in range(5)],)
+
+    @classmethod
+    def setUpClassWarehouse(cls):
+        cls.warehouse = cls.env.ref("stock.warehouse0")
+        cls.warehouse.write(
+            {
+                "cutoff_time": CUTOFF_TIME,
+                "apply_cutoff": True,
+                "calendar_id": cls.calendar,
+            }
+        )
+
+    @classmethod
+    def setUpClassPartner(cls):
+        cls.customer = cls.env.ref("base.res_partner_12")
+        cls.customer.order_delivery_cutoff_preference = "warehouse_cutoff"
+        cls.customer.delivery_time_preference = "workdays"
+
+    @classmethod
+    def setUpClassProduct(cls):
+        cls.product = cls.env.ref("product.product_product_3")
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestSaleOrderDates, cls).setUpClass()
+        cls.env = cls.env(
+            context=dict(cls.env.context, tracking_disable=True, tz="UTC")
+        )
+        cls.setupClassCompany()
+        cls.setUpClassCalendar()
+        cls.setUpClassWarehouse()
+        cls.setUpClassPartner()
+        cls.setUpClassProduct()
+
+    def _create_order(self, dt):
+        with freeze_time(dt.strftime("%Y-%m-%d %H:%M:%S")):
+            order = self.env["sale.order"].create(
+                {"partner_id": self.customer.id, "warehouse_id": self.warehouse.id}
+            )
+            self.env["sale.order.line"].create(
+                {
+                    "order_id": order.id,
+                    "product_id": self.product.id,
+                    "product_uom_qty": 1,
+                    "customer_lead": 1,
+                }
+            )
+            return order
+
+    @freeze_time(FRIDAY_BEFORE_CUTOFF)
+    def test_confirm_before_cutoff_last_weekday(self):
+        # - preparation is friday since sale_cutoff_time_delivery,
+        #   has not postponed the delivery
+        # - friday is compatible with the warehouse calendar.
+        # - delivery should be done on satursday
+        order = self._create_order(datetime.now())
```

Comment:
if you use datetime.now you are not freezing anything... :thinking:

Reply:
https://github.com/spulec/freezegun#context-manager
I think it's precisely the usecase, no ?

Reply:
![image](https://user-images.githubusercontent.com/13200247/133081183-4c3d37fd-60b7-43ad-9bde-3ec1bf21e48c.png)

Reply:
the point is that you are freezing the date w/ a dynamic date. You should pass a specific date to satisfy your test IMO.

Reply:
I'm using `freeze_time` as a decorator for the test methods, therefore, when calling `_create_order(datetime.now())`, the date is already frozen.
Probably not very efficient, I'm not sure I have to freeze the date in `_create_order`, but I'm pretty sure the date is frozen well.

Reply:
Why calling `freeze_time` in `_create_order` method, can't we rely only on the freeze done by the decorator on the test method?